keps/sig-node/127-usernamespaces-support/README.md

```diff
@@ -0,0 +1,810 @@
+that updates permissions and ownership of the files to be accesible by the
+`fsGroup` group ID). These pods should be run in `Host` mode.
+
+#### Container Runtime Support
+
+- **Docker**:
+  It only supports a [single ID
+  mappings](https://docs.docker.com/engine/security/userns-remap/) shared by all
+  containers running in the host. There is not support for [multiple ID
+  mappings](https://github.com/moby/moby/issues/28593) yet. Dockershim runtime
+  is only compatible with pods running in `Host` and `Cluster` modes. The user
+  has to guarantee that the ID mappings configured in Docker through the
+  `userns-remap` parameter and the cluster-wide range configured in the Kubelet
+  are the same. The dockershim implementation includes a check to verify that
+  the IDs mapping received from the Kubelet are equal to the ones configured in
+  Docker, returning an error otherwise.
+- **containerd**:
+  It's quite straigtforward to implement the CRI changes proposed below in
+  containerd/cri, we did it in
+  [this](https://github.com/kinvolk/containerd-cri/commits/mauricio/userns_poc)
+  PoC.
+- **cri-o**:
+  It recently [added](https://github.com/cri-o/cri-o/pull/3944) support for
+  user namespaces through a pod annotation. The extensions to make it work with
+  the current CRI changes are small.
+- TODO(Mauricio): gVisor, katacontainers?
+
+containerd and cri-o will provide support for the 3 possible values of `userNamespaceMode`.
+
+### Risks and Mitigations
+
+#### Breaking Existing Workloads
+
+Some features that could not work when the host user namespace is not shared are:
+
+- **Some Capabilities**:
+  The Linux kernel takes into consideration the user namespace a process is
+  running in while performing the capabilities check. There are some
+  capabilities that are only available in the root (host) user namespace such as
+  `CAP_SYS_TIME`, `CAP_SYS_MODULE` & `CAP_MKNOD`.
+
+- **Sharing Host Namespaces**:
+  There are some limitations in the Linux kernel and in the runtimes that
+  prevents sharing other host namespaces when the host user namespace is not
+  shared.
+  TODO(Mauricio): Put links to those limitations?
+
+In order to avoid breaking existing workloads `Host` is the default value of `userNamespaceMode`.
+
+#### Duplicated Snapshots of Container Images
+
+The owners of the files of a container image have to been set accordingly to the
+ID mappings used for that container. The runtimes perform a `chown` operation
+over the image snapshot when it's pulled. This presents a risk as it potentially
+increases the time and the storage needed to handle the container images.
+
+There is not immediate mitigation for it, [we
+talked](https://lists.linuxfoundation.org/pipermail/containers/2020-September/042230.html)
+to the Linux kernel community and [they
+replied](https://lists.linuxfoundation.org/pipermail/containers/2020-September/042230.html)
+they are working on a solution for it.
```

Comment:
You can add:
If the kernel provides a solution for this problem, that would be something that container runtime should use. It does not impact the Kubelet. Similarly, this does not impact the CRI gRPC spec.